packages/angular/ssr/src/routes/route-tree.ts

```diff
@@ -0,0 +1,224 @@
+    let lowestIndex = Infinity;
+
+    /**
+     * Recursively traverses the route tree to find the best match.
+     */
+    const traverse = (currentNode: RouteTreeNode, remainingSegments: string[]): void => {
+      const { insertionIndex, metadata, children } = currentNode;
+      if (!remainingSegments.length) {
+        // Reached a leaf node, check if it's a better match
+        if (metadata && insertionIndex < lowestIndex) {
+          bestMatchMetadata = metadata;
+          lowestIndex = insertionIndex;
+        }
+
+        return;
+      }
+
+      if (!children) {
+        return;
+      }
+
+      const [segment, ...restSegments] = remainingSegments;
+
+      // Try to match the exact segment
+      let childNode = children.get(segment);
+      if (childNode) {
+        traverse(childNode, restSegments);
+      }
+
+      // Try to match with a wildcard
+      childNode = children.get('*');
+      if (childNode) {
+        traverse(childNode, restSegments);
+      }
+    };
+
+    const segments = stripTrailingSlash(route).split('/');
+    traverse(this.root, segments);
+
+    return bestMatchMetadata;
+  }
+
+  /**
+   * Converts the route tree into a nested object representation.
+   * @returns A nested object where each key is a route segment and each value is an object containing
+   *          metadata and possibly further nested route segments.
+   */
+  toObject(): RouteNodeObjectRepresentation {
+    const result: RouteNodeObjectRepresentation = {};
+
+    /**
+     * Recursively traverses the route tree to build the nested object.
+     */
+    const traverse = (
+      node: RouteTreeNode,
+      currentPath: string,
+      currentObj: RouteNodeObjectRepresentation,
+    ) => {
+      if (node.metadata) {
+        currentObj.metadata = node.metadata;
+      }
+
+      if (node.children) {
+        for (const [segment, childNode] of node.children) {
+          currentObj.children ??= {};
+          currentObj.children[segment] = {};
+          traverse(childNode, joinUrlParts(currentPath, segment), currentObj.children[segment]);
+        }
+      }
+    };
+
+    traverse(this.root, '', result);
+
+    return result;
+  }
+
+  /**
+   * Constructs a RouteTree from a nested object representation.
+   * @param nestedObject A nested object where each key is a route segment and each value contains
+   *                     metadata and possibly further nested route segments.
+   * @returns A RouteTree constructed from the nested object.
+   */
+  static fromObject(nestedObject: RouteNodeObjectRepresentation): RouteTree {
+    const tree = new RouteTree();
+
+    /**
+     * Recursively traverses the nested object to insert routes into the tree.
+     */
+    const traverse = (currentObj: RouteNodeObjectRepresentation, currentPath: string) => {
+      if (currentObj.metadata) {
+        // Insert the current path with its metadata into the route tree
+        tree.insert(currentPath, currentObj.metadata);
```

Comment:
Question: How does this work with ordering? `tree.insert` is sensitive to ordering right? Is `RouteTree.fromObject(tree.toObject())` guaranteed to preserve ordering and have the same match behavior?

Even if that does work, wouldn't this imply that `RouteTree.fromObject` is sensitive to the ordering of properties in `children: { /* ... */ }`? That feels brittle to me and easy to not understand when using this API.

Reply:
> Question: How does ordering affect this process? Since `tree.insert` depends on ordering, does `RouteTree.fromObject(tree.toObject())` reliably maintain that order and ensure the same matching behavior?

The test results suggest that order is preserved.

> However, if this is the case, doesn't it imply that `RouteTree.fromObject` depends on the order of properties in children: { /* ... */ }? This seems fragile and could be confusing when using the API.

Yes, the function does rely on order, but isn't order also crucial in router configuration? In both cases, the sequence matters.

Would it be clearer to change the structure from:
```js
children: {
  '': {
    children: {
      first: {
        metadata: { route: '/first' },
      },
      '*': {
        metadata: { route: '/:id' },
      },
      second: {
        metadata: { route: '/second' },
      },
    },
  },
},
```

to 

```js
children: [
  {
    segment: '',
    children: [
      {
        segment: 'first',
        metadata: { route: '/first' },
      },
      {
        segment: '*',
        metadata: { route: '/:id' },
      },
      {
        segment: 'second',
        metadata: { route: '/second' },
      },
    ],
  },
]

```

Or perhaps a flattened version of the above?

```js
routes: [
  {
    path: '/something/first',
    metadata: { route: '/something/first' },
  },
  {
    path: '/something/[:id]',
    metadata: { route: '/something/[:id]' },
  },
  {
    path: '/something/second',
    metadata: { route: '/something/second' },
  },
]

```